tunnel/intra/doh/client_auth.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2020 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package doh
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"io"
+	"sync"
+)
+
+// CertificateLoader interface for requesting ClientAuth instances.
+type CertificateLoader interface {
+	// Request a ClientAuth instance (blocking).
+	// returns nil (no authentication) or a ClientAuth instance.
+	LoadClientCertificate() ClientAuth
+}
+
+// ClientAuth interface for providing TLS certificates and signatures.
+type ClientAuth interface {
+	// GetClientCertificate returns the client certificate (if any).
+	// It does not block or cause certificates to load.
+	// Returns a DER encoded X.509 client certificate.
+	GetClientCertificate() []byte
+	// GetIntermediateCertificate returns the chaining certificate (if any).
+	// It does not block or cause certificates to load.
+	// Returns a DER encoded X.509 certificate.
+	GetIntermediateCertificate() []byte
+	// Request a signature on a digest.
+	Sign(digest []byte) []byte
+}
+
+// clientAuthWrapper manages certificate loading and usage during TLS handshakes.
+// Implements crypto.Signer.
+type clientAuthWrapper struct {
+	sync.Mutex
+	loadCertificateOnce sync.Once
+	loader              CertificateLoader
+
+	certificate tls.Certificate
```

Comment:
Caching the certificate here makes this code somewhat more complicated (it makes this object stateful, and duplicates state that is also held in `loader`).  Is this just a performance optimization, or is there a reason (UX flow?) why this has to be cached?

Reply:
No, it's not a UX flow thing.  I don't know how performant bindings from Go to Java will be.  I've removed this caching as I agree that it's likely unnecessary.

Reply:
It looks like your change to remove the caching, and other changes to this code, went into a new, separate pull request.  Please close that pull request and put the changes in this one instead so we can maintain the review context.  This PR is from `SeanBurford:master`, so if you commit your changes to `master` and push `master` to `origin`, this PR should update automatically.